pennylane/templates/subroutines/semi_adder.py

```diff
@@ -0,0 +1,286 @@
+        ]
+
+    ck, tk, aux = wires
+
+    return [qml.CNOT([ck, tk]), qml.TemporaryAND([ck, tk, aux]), qml.CNOT([ck, aux])]
+
+
+def _right_operator(wires, ik_is_zero=False):
+    """Implement the right block in figure 2, https://arxiv.org/pdf/1709.06648"""
+
+    if not ik_is_zero:
+        ck, ik, tk, aux = wires
+        return [
+            qml.CNOT([ck, aux]),
+            qml.adjoint(qml.TemporaryAND([ik, tk, aux])),
+            qml.CNOT([ck, ik]),
+            qml.CNOT([ik, tk]),
+        ]
+
+    ck, tk, aux = wires
+
+    return [qml.CNOT([ck, aux]), qml.adjoint(qml.TemporaryAND([ck, tk, aux]))]
+
+
+class SemiAdder(Operation):
+    r"""This operator performs the plain addition of two integers :math:`x` and :math:`y` in the computational basis:
+
+    .. math::
+
+        \text{SemiAdder} |x \rangle | y \rangle = |x \rangle | x + y  \rangle,
+
+    This can be called a semi-out-place addition operation,
+    or more specifically, the operation is a quantum-quantum in-place addition.
+
+    The implementation is based on `arXiv:1709.06648 <https://arxiv.org/abs/1709.06648>`_.
+
+    Args:
+        x_wires (Sequence[int]): the wires that store the integer :math:`x`
+        y_wires (Sequence[int]): the wires that store the integer :math:`y`
+        work_wires (Sequence[int]): the auxiliary wires to use for the addition. ``len(y_wires) - 1`` work
+            wires should be provided.
+
+    **Example**
+
+    This example computes the sum of two integers :math:`x=3` and :math:`y=4`.
+
+    .. code-block::
+
+        x = 3
+        y = 4
+
+        wires = qml.registers({"x":3, "y":6, "work":5})
+
+        dev = qml.device("default.qubit", shots=1)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisEmbedding(x, wires=wires["x"])
+            qml.BasisEmbedding(y, wires=wires["y"])
+            qml.SemiAdder(wires["x"], wires["y"], wires["work"])
+            return qml.sample(wires=wires["y"])
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [0 0 0 1 1 1]
+
+    The result :math:`[0 0 0 1 1 1]`, is the binary representation of :math:`3 + 4 = 7`.
+
+    Note that the result is subject to the size of the ``y_wires`` register, since due to overhead
+    it is taken modulo :math:`2^{\text{len(y_wires)}}`, as shown in this example:
+
+    .. code-block::
+
+        x = 3
+        y = 1
+
+        wires = qml.registers({"x":3, "y":2, "work":1})
+
+        dev = qml.device("default.qubit", shots=1)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisEmbedding(x, wires=wires["x"])
+            qml.BasisEmbedding(y, wires=wires["y"])
+            qml.SemiAdder(wires["x"], wires["y"], wires["work"])
+            return qml.sample(wires=wires["y"])
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [0 0]
+
+    The result :math:`[0\ 0]` is the binary representation of :math:`3 + 1 = 4 \equiv 0 \mod 2^2`.
+
+    .. details::
+        :title: Usage Details
```

Comment:
[Optional] Does it really need a usage details section? You might just mention this in the Args description, e.g., 

x_wires (Sequence[int]): The wires that store the integer :math:`x`.  Must contain at least :math:`\lceil \log_2(x)\rceil` wires to represent :math:`x`.

And the same for others.

Reply:
Nice suggestion 👍